### DISCUSSION TITLE
[Detector Support]: OpenVino configured but erroring when switching to GPU

### DISCUSSION THREAD

**Discussion by Samywamy10** [blakeblackshear/frigate#12012] [Detector Support] [answered]
### Describe the problem you are having

Trying to fix ~150ms inference time.

Using the latest 0.14 beta although I suspect this issue still exists in ver 0.13 given the [inference speed of 247ms in my previous issue](https://github.com/blakeblackshear/frigate/discussions/11676)

When I switch to the GPU detector, its unable to initialise with the error `[GPU] Context was not initialized for 0 device`

My setup is using a Beelink S12 with an Intel N100 processor in a docker LXC in Proxmox. Originally I thought the issue was because openvino couldn't detect the GPU so I spent a bucketload of time trying to get OpenCL working, and finally thanks to [this post](https://github.com/tteck/Proxmox/discussions/2928#discussioncomment-9344576) I got it going.

So now, when I run `clinfo` I get:

```
root@docker-frigate:/opt/frigate/config# clinfo
Number of platforms                               1
  Platform Name                                   Intel(R) OpenCL Graphics
....
```

I can also see the GPU when I run the [hello query device script](https://docs.openvino.ai/2024/learn-openvino/openvino-samples/hello-query-device.html)

```
[ INFO ] GPU :
[ INFO ]        SUPPORTED_PROPERTIES:
[ INFO ]                AVAILABLE_DEVICES: 0
[ INFO ]                RANGE_FOR_ASYNC_INFER_REQUESTS: 1, 2, 1
[ INFO ]                RANGE_FOR_STREAMS: 1, 2
[ INFO ]                OPTIMAL_BATCH_SIZE: 1
[ INFO ]                MAX_BATCH_SIZE: 1
[ INFO ]                DEVICE_ARCHITECTURE: GPU: vendor=0x8086 arch=v769.0.0
[ INFO ]                FULL_DEVICE_NAME: Intel(R) UHD Graphics (iGPU)
[ INFO ]                DEVICE_UUID: 8680d146000000000002000000000000
[ INFO ]                DEVICE_LUID: 409a0000499a0000
[ INFO ]                DEVICE_TYPE: Type.INTEGRATED
[ INFO ]                DEVICE_GOPS: {<Type: 'float16'>: 576.0, <Type: 'float32'>: 288.0, <Type: 'int8_t'>: 1152.0, <Type: 'uint8_t'>: 1152.0}
[ INFO ]                OPTIMIZATION_CAPABILITIES: FP32, BIN, FP16, INT8, EXPORT_IMPORT
[ INFO ]                GPU_DEVICE_TOTAL_MEM_SIZE: 15076028416
[ INFO ]                GPU_UARCH_VERSION: 769.0.0
[ INFO ]                GPU_EXECUTION_UNITS_COUNT: 24
[ INFO ]                GPU_MEMORY_STATISTICS: {}
[ INFO ]                PERF_COUNT: False
[ INFO ]                MODEL_PRIORITY: Priority.MEDIUM
[ INFO ]                GPU_HOST_TASK_PRIORITY: Priority.MEDIUM
[ INFO ]                GPU_QUEUE_PRIORITY: Priority.MEDIUM
[ INFO ]                GPU_QUEUE_THROTTLE: Priority.MEDIUM
[ INFO ]                GPU_ENABLE_LOOP_UNROLLING: True
[ INFO ]                GPU_DISABLE_WINOGRAD_CONVOLUTION: False
[ INFO ]                CACHE_DIR: 
[ INFO ]                CACHE_MODE: CacheMode.OPTIMIZE_SPEED
[ INFO ]                PERFORMANCE_HINT: PerformanceMode.LATENCY
[ INFO ]                EXECUTION_MODE_HINT: ExecutionMode.PERFORMANCE
[ INFO ]                COMPILATION_NUM_THREADS: 4
[ INFO ]                NUM_STREAMS: 1
[ INFO ]                PERFORMANCE_HINT_NUM_REQUESTS: 0
[ INFO ]                INFERENCE_PRECISION_HINT: <Type: 'float16'>
[ INFO ]                ENABLE_CPU_PINNING: False
[ INFO ]                DEVICE_ID: 0
```

Yet still getting the error when Frigate starts. Any ideas?

### Version

0.14.0-4e5a6eb

### Frigate config file

```yaml
ffmpeg:
  hwaccel_args: preset-vaapi

detectors:
  ov:
    type: openvino
    device: GPU # this is what doesn't work. CPU works with ~150ms inference time

model:
  width: 300
  height: 300
  input_tensor: nhwc
  input_pixel_format: bgr
  path: /openvino-model/ssdlite_mobilenet_v2.xml
  labelmap_path: /openvino-model/coco_91cl_bkgr.txt
```

### docker-compose file or Docker CLI command

```yaml
version: '3.9'

services:

  frigate:
    container_name: frigate
    privileged: true
    restart: unless-stopped
    image: ghcr.io/blakeblackshear/frigate:0.14.0-beta3
    shm_size: "512mb" # update for your cameras based on calculation above
    devices:
      - /dev/bus/usb:/dev/bus/usb
      - /dev/dri/renderD128 # for intel hwaccel, needs to be updated for your hardware
    volumes:
      - /etc/localtime:/etc/localtime
      - /opt/frigate/config:/config
      - /cctv_clips:/media/frigate
      - type: tmpfs # Optional: 1GB of memory, reduces SSD/SD Card wear
        target: /tmp/cache
        tmpfs:
          size: 1000000000
    ports:
      - "5000:5000"
      - "1935:1935" # RTMP feeds
```

### Relevant log output

```shell
2024-06-17 20:52:47.516588921  [2024-06-17 20:52:47] detector.ov                    INFO    : Starting detection process: 370
2024-06-17 20:52:47.708147481  Process detector:ov:
2024-06-17 20:52:47.709877527  Traceback (most recent call last):
2024-06-17 20:52:47.710224425    File "/usr/lib/python3.9/multiprocessing/process.py", line 315, in _bootstrap
2024-06-17 20:52:47.710227486      self.run()
2024-06-17 20:52:47.710490307    File "/usr/lib/python3.9/multiprocessing/process.py", line 108, in run
2024-06-17 20:52:47.710492604      self._target(*self._args, **self._kwargs)
2024-06-17 20:52:47.710742225    File "/opt/frigate/frigate/object_detection.py", line 102, in run_detector
2024-06-17 20:52:47.710744712      object_detector = LocalObjectDetector(detector_config=detector_config)
2024-06-17 20:52:47.710948189    File "/opt/frigate/frigate/object_detection.py", line 53, in __init__
2024-06-17 20:52:47.710950071      self.detect_api = create_detector(detector_config)
2024-06-17 20:52:47.711154462    File "/opt/frigate/frigate/detectors/__init__.py", line 18, in create_detector
2024-06-17 20:52:47.711156131      return api(detector_config)
2024-06-17 20:52:47.711345388    File "/opt/frigate/frigate/detectors/plugins/openvino.py", line 38, in __init__
2024-06-17 20:52:47.711346989      self.interpreter = self.ov_core.compile_model(
2024-06-17 20:52:47.711566779    File "/usr/local/lib/python3.9/dist-packages/openvino/runtime/ie_api.py", line 521, in compile_model
2024-06-17 20:52:47.711568644      super().compile_model(model, device_name, {} if config is None else config),
2024-06-17 20:52:47.711771082  RuntimeError: Exception from src/inference/src/cpp/core.cpp:123:
2024-06-17 20:52:47.711772836  Exception from src/inference/src/dev/plugin.cpp:59:
2024-06-17 20:52:47.711773864  Check 'contexts.count(device_id)' failed at src/plugins/intel_gpu/src/plugin/plugin.cpp:204:
2024-06-17 20:52:47.711776681  [GPU] Context was not initialized for 0 device
```

### Operating system

Proxmox

### Install method

Docker Compose

### Object Detector

OpenVino

**Comment by NickM-27** [comment]
Are you sure the GPU is available in the container? Easiest is to run vainfo (hardware info) inside of Frigate ui

**Reply by Samywamy10** [reply]
This is the output of hardware info

```

Vainfo Output
Return Code: 0

Process Output:

vainfo: VA-API version: 1.17 (libva 2.10.0)
vainfo: Driver version: Intel iHD driver for Intel(R) Gen Graphics - 23.1.1 ()
vainfo: Supported profile and entrypoints
VAProfileNone : VAEntrypointVideoProc
VAProfileNone : VAEntrypointStats
VAProfileMPEG2Simple : VAEntrypointVLD
VAProfileMPEG2Simple : VAEntrypointEncSlice
VAProfileMPEG2Main : VAEntrypointVLD
VAProfileMPEG2Main : VAEntrypointEncSlice
VAProfileH264Main : VAEntrypointVLD
VAProfileH264Main : VAEntrypointEncSlice
```


I've also ran both `clinfo` and the hello query device script and both can't see the GPU. Currently investigating this. Seems like it has access to `/dev/dri/renderD128`...maybe a driver issue? None of the guides to do with OpenVino or Frigate within Docker seem to say anything about needing to install drivers though. Will test

**Reply by NickM-27** [reply]
this shows that iHD is working correctly. Might want to make sure /dev/dri/card0 is also passed in. preset-vaapi for hwaccel_args should work, would be good to confirm that

**Reply by Samywamy10** [reply, marked as answer]
Ahhh so just ran these same [commands from here](https://github.com/tteck/Proxmox/discussions/2928#discussioncomment-9344576) WITHIN the docker container and it seems to now detect the GPU properly.

It actually error on a few of the things it was installing but one of them seemed to do the trick. Seems like a driver was missing. I don't know enough about how GPU passthrough is supposed to work in Docker to know what the solution is or where this driver is supposed to come from

Anyway, changed the detector type back to `GPU` and the container starts without error.

Inference speed a ~juicy 35ms.....~ averaging about 85ms now

**Reply by Samywamy10** [reply]
Also can confirm that ffmpeg is correctly using the igpu - I can see load from ffmpeg when doing `intel_gpu_top`. It was just Openvino specifically that seems to need OpenCL drivers which were missing from both the proxmox LXC as well as the docker instance.

(screenshot taken afterwards but beforehand it was just the ffmepg processes)

![CleanShot 2024-06-17 at 23 17 47@2x](https://github.com/blakeblackshear/frigate/assets/1314053/d8d108ca-1c88-4a15-b5a5-78e08ce978da)

**Reply by Samywamy10** [reply]
Another data point, vainfo when correct drivers installed is significantly longer

```
Vainfo Output
Return Code: 0

Process Output:

vainfo: VA-API version: 1.17 (libva 2.10.0)
vainfo: Driver version: Intel iHD driver for Intel(R) Gen Graphics - 23.1.1 ()
vainfo: Supported profile and entrypoints
VAProfileNone : VAEntrypointVideoProc
VAProfileNone : VAEntrypointStats
VAProfileMPEG2Simple : VAEntrypointVLD
VAProfileMPEG2Simple : VAEntrypointEncSlice
VAProfileMPEG2Main : VAEntrypointVLD
VAProfileMPEG2Main : VAEntrypointEncSlice
VAProfileH264Main : VAEntrypointVLD
VAProfileH264Main : VAEntrypointEncSlice
VAProfileH264Main : VAEntrypointFEI
VAProfileH264Main : VAEntrypointEncSliceLP
VAProfileH264High : VAEntrypointVLD
VAProfileH264High : VAEntrypointEncSlice
VAProfileH264High : VAEntrypointFEI
VAProfileH264High : VAEntrypointEncSliceLP
VAProfileVC1Simple : VAEntrypointVLD
VAProfileVC1Main : VAEntrypointVLD
VAProfileVC1Advanced : VAEntrypointVLD
VAProfileJPEGBaseline : VAEntrypointVLD
VAProfileJPEGBaseline : VAEntrypointEncPicture
VAProfileH264ConstrainedBaseline: VAEntrypointVLD
VAProfileH264ConstrainedBaseline: VAEntrypointEncSlice
VAProfileH264ConstrainedBaseline: VAEntrypointFEI
VAProfileH264ConstrainedBaseline: VAEntrypointEncSliceLP
VAProfileVP8Version0_3 : VAEntrypointVLD
VAProfileHEVCMain : VAEntrypointVLD
VAProfileHEVCMain : VAEntrypointEncSlice
VAProfileHEVCMain : VAEntrypointFEI
VAProfileHEVCMain : VAEntrypointEncSliceLP
VAProfileHEVCMain10 : VAEntrypointVLD
VAProfileHEVCMain10 : VAEntrypointEncSlice
VAProfileHEVCMain10 : VAEntrypointEncSliceLP
VAProfileVP9Profile0 : VAEntrypointVLD
VAProfileVP9Profile0 : VAEntrypointEncSliceLP
VAProfileVP9Profile1 : VAEntrypointVLD
VAProfileVP9Profile1 : VAEntrypointEncSliceLP
VAProfileVP9Profile2 : VAEntrypointVLD
VAProfileVP9Profile2 : VAEntrypointEncSliceLP
VAProfileVP9Profile3 : VAEntrypointVLD
VAProfileVP9Profile3 : VAEntrypointEncSliceLP
VAProfileHEVCMain12 : VAEntrypointVLD
VAProfileHEVCMain12 : VAEntrypointEncSlice
VAProfileHEVCMain422_10 : VAEntrypointVLD
VAProfileHEVCMain422_10 : VAEntrypointEncSlice
VAProfileHEVCMain422_12 : VAEntrypointVLD
VAProfileHEVCMain422_12 : VAEntrypointEncSlice
VAProfileHEVCMain444 : VAEntrypointVLD
VAProfileHEVCMain444 : VAEntrypointEncSliceLP
VAProfileHEVCMain444_10 : VAEntrypointVLD
VAProfileHEVCMain444_10 : VAEntrypointEncSliceLP
VAProfileHEVCMain444_12 : VAEntrypointVLD
VAProfileHEVCSccMain : VAEntrypointVLD
VAProfileHEVCSccMain : VAEntrypointEncSliceLP
VAProfileHEVCSccMain10 : VAEntrypointVLD
VAProfileHEVCSccMain10 : VAEntrypointEncSliceLP
VAProfileHEVCSccMain444 : VAEntrypointVLD
VAProfileHEVCSccMain444 : VAEntrypointEncSliceLP
VAProfileAV1Profile0 : VAEntrypointVLD
VAProfileHEVCSccMain444_10 : VAEntrypointVLD
VAProfileHEVCSccMain444_10 : VAEntrypointEncSliceLP
```

**Reply by Samywamy10** [reply]
Ultimately I believe this all to be a problem with the Proxmox kernel version. See my [post here](https://github.com/tteck/Proxmox/discussions/2928#discussioncomment-9812253), updating it helped :)

**Reply by klingon888** [reply]
> Ultimately I believe this all to be a problem with the Proxmox kernel version. See my [post here](https://github.com/tteck/Proxmox/discussions/2928#discussioncomment-9812253), updating it helped :)

Thanks for posting this. I had the same error and can confirm that updating Proxmox kernel to 6.8.8-1 fixes the Openvino errors. I'm also on Intel N100 installed using tteck's scripts. The inference speed took a few mins to settle but once settled, its now at 12ms. GPU usage idle at 5.5% with 4 cams.

**Reply by Redsandro** [reply]
@Samywamy10 commented:
> Ahhh so just ran these same [commands from here](https://github.com/tteck/Proxmox/discussions/2928#discussioncomment-9344576) WITHIN the docker container and it seems to now detect the GPU properly.
> 
> It actually error on a few of the things it was installing but one of them seemed to do the trick. 

Thank you! I spent so much time on all these tutorials on getting to forward my iGPU, but it was probably fine hours ago. This did the trick! The Frigate docker image just needs a couple of these packages installed.

I get a bunch of errors too, but it works now.

Is this already tracked on the issue tracker? (**edit:** https://github.com/blakeblackshear/frigate/discussions/12260) Because once you do the wrong kind of restart, or docker down docker up, you have to do this all over again. It is not complicated, but not nice either, because you need to change your config temporary to stop the container from crashing in a loop so you can download and install the packages.

For reference; my [specs](https://github.com/blakeblackshear/frigate/discussions/12238) are different:

- ASUS NUC 14 Pro Plus
- Intel Core Ultra 
- Intel Arc Alchemist iGPU

**Reply by philippedegeer** [reply]
@Samywamy. Same hardware as you have (Core Ultra 7 with Arc iGPU). Proxmox 6.8.8-3 did not solve the Openvino iGPU issue. Had to run the commands to update the Intel Compute Runtime.
